### PR TITLE
build:adds subtask emission-disperse-bal

### PR DESCRIPTION
**Pull request by doncesarts**
###  adds subtask emission-disperse-bal 

1. it gets the report from the repo https://github.com/balancer-labs/bal-mining-scripts/blob/master/report
2. scale the values to match with the MTA disperse forwarder  balance
3. disperse the tokens to the recipients 
4. adds debug package to the project

### CODE REVIEW

**Comment by doncesarts** [comment]
Not quite sure if these changes must be part of the tests , or a version where the http request is mock, but it sets the basis for a test of the subtask 

```
import { expect } from "chai"
import { simpleToExactAmount } from "@utils/math"
import { assertBNClose } from "@utils/assertions"
import { getBalancerPolygonReport } from "../../tasks/utils/emission-disperse-bal"

describe("disperse balancer rewards", () => {
    it("gets report of a valid week", async () => {
        // Balance with 18 decimals , ie "0.25028450" to 250284500000000000
        const disperseForwarderBalance = simpleToExactAmount(150000)

        const { disperser, balancer } = await getBalancerPolygonReport(79, disperseForwarderBalance)

        console.log("disperse forwarder balance", disperseForwarderBalance.toString())
        console.log("disperse rewards total", disperser.total.toString())
        console.log("disperse recipients total", disperser.recipients.length)
        console.log("disperse recipients total", disperser.values.length)

        console.log("balancer rewards count", balancer.rewards.length)
        console.log("balancer rewards total", balancer.total.toString())
        assertBNClose(disperseForwarderBalance, disperser.total, 1000)

        // expect(disperseForwarderBalance, "disperse balance").to.eq(disperser.total)

        expect(balancer.rewards.length, "disperse recipients rewards count").to.eq(disperser.recipients.length)
        expect(balancer.rewards.length, "disperse values rewards count").to.eq(disperser.values.length)
    })
    it("fails getting the report of an invalid week", () => {
        const disperseForwarderBalance = simpleToExactAmount(150000)

        const promise = getBalancerPolygonReport(-79, disperseForwarderBalance)
        promise.catch((reason) => expect(reason.message).to.eq("404: Not Found"))
    })
})

```

File: tasks/utils/emission-disperse-bal.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import https from "https"
+import { BN, simpleToExactAmount } from "@utils/math"
+
+const MTA_ADDRESS = "0xF501dd45a1198C2E1b5aEF5314A68B9006D842E0"
```
Comment (naddison36) [inline tasks/utils/emission-disperse-bal.ts:4]: You can get the address from `PMTA.address`

File: tasks/utils/emission-disperse-bal.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    total: BN
+    rewards: Array<BalancerReward>
+}
+
+interface BalancerPolygonReport {
+    disperser: DisperseDetails
+    balancer: BalancerDetails
+}
+
+const log = (message: string, debugMode = false): void => {
```
Comment (naddison36) [inline tasks/utils/emission-disperse-bal.ts:26]: Implementing a debugger has been on my TODO list for a while. 
I usually use `debug` https://www.npmjs.com/package/debug in my other projects but I'm open to other suggestions.

File: tasks/utils/emission-disperse-bal.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+export const fetchBalancerReport = async (
+    report: number,
+    options = { debugMode: false },
+): Promise<Array<{ address: string; amount: string }>> => {
+    const url = `${REPORT_URL}/${report}/__polygon_${MTA_ADDRESS}.json`
+    log(`fetchBalancerReport: get balancer-labs for report ${report}`, options.debugMode)
+    log(`fetchBalancerReport: download report from url :${url}`, options.debugMode)
+
+    return new Promise((resolve, reject) => {
+        https.get(url, (res) => {
```
Comment (naddison36) [inline tasks/utils/emission-disperse-bal.ts:47]: we do have `axios` in the dependencies. That might be easier to use. eg https://github.com/mstable/mStable-contracts/blob/5d57eaecacc0deeb37ecc7b7b6137314da7aed81/tasks/stakingV1.ts#L21

File: tasks/utils/emission-disperse-bal.ts (hunk 1 above)
Comment (doncesarts) [reply tasks/utils/emission-disperse-bal.ts]: resolved with e270e136bd3e196c30033659347192d30ff84da0

File: tasks/utils/emission-disperse-bal.ts (hunk 2 above)
Comment (doncesarts) [reply tasks/utils/emission-disperse-bal.ts]: resolved with e270e136bd3e196c30033659347192d30ff84da0

File: tasks/utils/emission-disperse-bal.ts (hunk 3 above)
Comment (doncesarts) [reply tasks/utils/emission-disperse-bal.ts]: resolved with e270e136bd3e196c30033659347192d30ff84da0

**Comment by doncesarts** [comment]
Hi @naddison36  in order to see the debug output a node variable must be set as follows:
```
DEBUG=mstable* 
```

The next task that we can add in the backlog is to add that into `logTxDetails` utility , add documentation about it , update any CI Build to pass this new env variable.

File: tasks/emissions.ts [hunk 4]
```diff
@@ -53,21 +55,20 @@ subtask("emission-disperse-bal", "Disperse Polygon Balancer Pool MTA rewards in
...
-        const recipients = []
-        const values = []
-        const tx = await disperseForwarder.disperseToken(recipients, values)
-        await logTxDetails(tx, `Disperse Balancer Pool MTA rewards to ${recipients.length} recipients`)
+        try {
+            // Get the proportion the MTA balance in the DisperseForwarder contract to the recipients based off the bal-mining-script report.
+            const { disperser } = await getBalancerPolygonReport(taskArgs.report, mtaBalance)
+            const tx = await disperseForwarder.disperseToken(disperser.recipients, disperser.values)
+            await logTxDetails(tx, `Disperse Balancer Pool MTA rewards ${disperser.total}  to ${disperser.recipients} recipients`)
+        } catch (error) {
```
Comment (naddison36) [inline tasks/emissions.ts:68]: Nice error handling. I usually don't bother with task but I should.

File: tasks/utils/deploy-utils.ts [hunk 5]
```diff
@@ -25,6 +26,9 @@ export const deployContract = async <T extends Contract>(
     return contract
 }
 
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export const logger = (...args: string[]) => debug(`mstable:task:${args.join(":")}`)
```
Comment (naddison36) [inline tasks/utils/deploy-utils.ts:30]: we can move this into its own `tasks/utils/logger.ts` file.

Passing the the args and joining with `:` is a cool idea. I like

**Review by naddison36 — APPROVED** [review]
this is looking good.

**Pull request merged**